main/main_timer_sync.cpp

```diff
@@ -143,9 +166,15 @@ MainFrameTime MainTimerSync::advance_core(float p_frame_slice, int p_iterations_
 
 // calls advance_core, keeps track of deficit it adds to animaption_step, make sure the deficit sum stays close to zero
 MainFrameTime MainTimerSync::advance_checked(float p_frame_slice, int p_iterations_per_second, float p_idle_step) {
+	if (p_idle_step <= 0) {
+		WARN_PRINT_ONCE("p_idle_step not positive");
+	}
+
 	if (fixed_fps != -1)
 		p_idle_step = 1.0 / fixed_fps;
 
+	const float min_output_step = p_idle_step > 0 ? p_idle_step * .25 : 1E-6;
```

Comment:
Shouldn't the minimum just be zero? Also, maybe this line should be moved down to where the if statement is

Reply:
If we think that _process() should be able to handle zero as delta argument, yes, min_output_step should just be zero. The unexpectedly crashing code from #26887 would handle it fine.
 In the current form, it probably shouldn't be moved further down because the next line changes p_idle_step, possibly making it negative; that would then lead to the minimal minimal step of 1E-6, which I'm not overly fond of.

I count that as two votes in favor of allowing delta=0 as argument to _process. Anyone against it?

Reply:
```
func _process(delta):
    translation.x += 100.0 * delta # Delta being 0 is valid
```

Reply:
> Everyone expects delta to be change-in-time-since-last-process. It is physically impossible for a perfectly accurate timer to say no time has passed, as calling functions always takes time

It is physically impossible to store a delta time less than the float epsilon, and I would vote in favor of more accurate long-term timers. If you add `delta` to a time variable every frame, and the script is executing faster than `1/epsilon` frames per second, then the timer would run too fast. Regardless, one would need hundreds of thousands of FPS to get a zero delta in the first place (possibly millions).

Reply:
For the timers Godot uses, granularity is bigger than the CPU clock (I think it was 10^6 ticks per second?). So yes, it is physically possible to get a measurement of zero for the passed time since the last frame, however improbable that is. If you're dividing by the delta argument to update, you're making two mistakes: Doing a potentially failing operation without checking, and doing a numerically unstable operation for small delta arguments. Yes, that these error sources exist is a surprise for novices. Ideally, we'd throw a delta=0 and delta=10^-6 argument at them in debug mode every once in a while to make everyone get used to the possibility; it's hard do so that without disrupting game flow, however.

Reply:
I don't see what all the fuss is. No gain will come from receiving a time based event with no time having passed. I for one cannot believe that this simple thing is so debated and left this way.

I have written toy engines myself in C and GL and have had to fix jitter and the like, and none of it was ever because no time passed between frames. This sounds 100% like covering up the real problem: Bad time source.

On Win32, QueryPerformanceCounter provides for nanosecond accuracy on most hardware that I tested. This guarantees that the second query will be after the first, and there's no way that, given the amount of work Godot must do, for a delta of 0 or less to EVER appear.

Even if it did appear, you're doing work based on time. Whatever that work is, you'd do none of it if delta was zero. Wasted cycles, wasted energy. Just throw out the process event if <= 0, seriously. It's not like anything is going to break... no time passed, remember? And the next tiny slice of time will get the process call, so again... no problems, but it's fixed. No potential gachas for newbies who (imo correctly) assume that time always marches forward. No chance of causing any issues because of a dropped _process, because it'll happen again *extremely* soon.

Reply:
You are 100% correct: All of this IS a way to compensate for a bad timer source. The problem isn't the timer resolution, though, it's that the time is taken on the CPU. In the best case (single threaded operation), when all the commands have been passed to the GPU. In the worst case (multithread), when all the scene data has been passed to the rendering thread. If you want stable frame time measurements, you should take the time when the GPU is starting to deliver the newly rendered frame to the screen. Alas, in GLES, that seems to be impossible. In Desktop OpenGL and Vulkan, you can use fences and queries to determine when the GPU flipped the buffers N frames ago. It's a bit difficult because it's asynchronous, but doable. Patches probably welcome.

The entire purpose of the file that is getting patched here is to take the imperfect timer measurements and derive stable process calls. Focus is on getting consistent number of physics steps per render frame. Without the code in this file, Godot would sometimes, even in perfect 60FPS VSync locked conditions, do jittery updates where each frame, anything between 0 and 2 physics updates were performed, because the imperfect timer measurements would lie just a little bit on either side of a threshold. Averaging over N frames makes these occurrences rarer, but does not fix them.

The negative timesteps that are corrected in this patch aren't coming from the timer itself, they're a bad output of the correction. Essentially, it's overshooting a bit one frame, then tries to backpedal. Maybe it would be possible to just prevent the overshooting; I don't think so, because that's what the system has to do to function correctly when frame times decrease. I opted for the post-correction also because that guarantees there are no side effects, it's easier to prove that the result will be correct, and it'll stay correct even if the core gets changed.

Just not calling _process() if the argument would be 0 is also not such a good option. There are three possible expectations here:
1. The sum of delta arguments to _process() is as close to the total program runtime as possible. Networked games rely on that.
2. the argument to _process() should be positive. That is what you want, even though it was never really guaranteed (just extremely unlikely to be false).
3. _process is called for every rendered frame. That is [documented](https://docs.godotengine.org/en/stable/getting_started/step_by_step/scripting_continued.html?highlight=_process#processing).
Best possible conformance with expectation 1, in the situation covered here, requires to either violate 2 or 3. You get screamed at by someone no matter what :)

The current, updated patch opts to violate expectation 1 a little to save 2 and 3. But not more than it is allowed to; the limit on real time/game time deviation still is less than physics_jitter_fix/physics_fps. I personally would prefer to break bad client code, but do so early; I don't really see a way to do that well, though, and don't want that to be the hill this bugfix dies on.

Reply:
> The negative timesteps that are corrected in this patch aren't coming from the timer itself, they're a bad output of the correction. Essentially, it's overshooting a bit one frame, then tries to backpedal. Maybe it would be possible to just prevent the overshooting; I don't think so, because that's what the system has to do to function correctly when frame times decrease. I opted for the post-correction also because that guarantees there are no side effects, it's easier to prove that the result will be correct, and it'll stay correct even if the core gets changed.

I'll just comment here, to show my support.  But the above comment is really is bang on.  This really needs to get clamped in all circumstances and no circumstance pass a negative delta to an iteration step.  If the timing PURELY used the jittery CPU timer sampler, it sill would guarantee >= delta times.  If any 'smoothing' or 'physics fix' post timer fix is causing a negative delta, this needs to get clamped to > 0 as soon as possible.  

With regards to violation of expectation 1, it should be ok if there is small amount of accumulated delta drift across frame/iteration boundaries, but not across the long-term time accumulator itself.  Which is think is covered here.  It's correc that we can't have the delta arguments to _process() continuously drift away from total program runtime.  Variance between iterations is fine, but long term definitely not fine.  As far as I can tell from this PR, all time is accounted for and accumulated in the end.